014-single-codebase.md

```diff
@@ -0,0 +1,68 @@
+# Single Codebase
+
+We should consolidate our services into a single repository, with a consolidated monolithic backend.
+
+**Discussion:** https://github.com/DoSomething/rfcs/pull/14
+
+## Problem
+
+We rebuilt our old [Drupal 7 monolith](https://github.com/dosomething/website-2013) into a suite of services – [Northstar](https://github.com/DoSomething/northstar), [Rogue](https://github.com/DoSomething/rogue), [Phoenix](https://github.com/DoSomething/phoenix-next), etc. – between 2013-2019. We avoided [the quagmire of a total rewrite](https://www.joelonsoftware.com/2000/04/06/things-you-should-never-do-part-i/) by extracting functionality [piece-by-piece](https://martinfowler.com/bliki/StranglerFigApplication.html) & have been generally happy with what we've built since then.
+
+However, breaking our monolith into a service-oriented architecture introduced new issues of its own:
+
+- **Blurred boundaries:** The [bounded contexts](https://martinfowler.com/bliki/BoundedContext.html) for each of our services (especially our identity & activity APIs) can be blurry. For example, is joining a club part of a user's identity or an action that they take? We're often forced to make this decision early in a feature's development process, before the technical work has fully taken shape.
+
+- **Technical overhead:** Logic that straddles multiple domains is also inevitably more complicated, requiring network requests for what could otherwise be a function call or `JOIN` statement. It makes traditionally simple logic into distributed systems problems where we have to decide whether we'll trade consistency for performance or vice versa.
+
+- **Testing limitations:** We know we get the best bang for our buck from [high-level integration tests](https://github.com/DoSomething/rfcs/blob/master/004-testing.md), but separating into services forces us to use mocking extensively because the majority of our application (e.g. users in Rogue, or users & posts in Phoenix) is always an external dependency in any given test! This places a large burden on our intra-service contracts.
+
+- **Knowledge silos:** Finally, this separation reinforces silos (e.g. "you have to talk to Dave about that app" or "that's X application's responsibility, not ours"). We have a smaller team today than when we first started splitting out into a service oriented architecture, and it's important that we can all understand & comfortably work on the whole system.
+
+
+
+## Proposal
+
+We should consolidate our backend services (Northstar, Rogue, and Chompy) into **a single monolithic Laravel application.**
```

Comment:
How would this work for the Rogue SQL and Northstar MongoDB databases?  Are you thinking they'd be separate, or migrate one into the other?

Reply:
For now, I think we'd keep both database connections and specify that the user's model is located on the MongoDB database via [Eloquent's `$collections` property](https://laravel.com/docs/8.x/eloquent#database-connection). Longer-term, I do think it's worth re-evaluating whether MongoDB is worth the overhead it incurs (in terms of technical onboarding & operational complexity) but that's probably another RFC!

Reply:
I think that makes sense. Perhaps we should spell it out, or would that make more sense to hold off for a formal technical spec? Chompy has a SQL database used primarily for the RTV import, that I'd imagine would be cleanest to migrate over to the Rogue DB.

Reply:
I'm starting to sort out a spec for how we'd tackle this work in practice, and will definitely include that there!

Reply:
I'm sure you're already considering this @DFurnes and I'm playing catch up after being gone for ~4 months... but thoughts on PostgreSQL for the monolith DB? It would support a monolith with services that used a SQL database and might be easier, down the road, to transition MongoDB data into it thanks to it's support for JSON-like data?

Reply:
I think that's a great idea! PostgreSQL seems like a really good fit with its JSON column support, and we could consider using Aurora Serverless to cut down costs a bit more along the way. Plus, we already have a lot of in-house expertise using PostgreSQL on the data side of the house! 🤓